# Sparse Optimization with Lasso Regression
# Feature Selection for High-Dimensional Data

## Javier Fernández Ramos

This notebook demonstrates how to use Lasso regression for feature selection in high-dimensional datasets. Lasso (Least Absolute Shrinkage and Selection Operator) is a linear regression technique that includes a regularization term to promote sparsity in the model coefficients, effectively selecting a subset of features.

The data was acquired by the public kaggle dataset "TCGA - LUSC | Lung Cancer Gene Expression Dataset" from https://www.kaggle.com/datasets/noepinefrin/tcga-lusc-lung-cell-squamous-carcinoma-gene-exp/data . 

Dataset consists of 551 patients, samples, each with 56970 differetn transcripts (expressed genes). 

Lung Cell Squamos Carcinoma is a cancer that occurs in lungs. Detecting and predicting it by Machine Learning is a clear challenge that could be very helpful.

## Preparing Data and Preprocessing

## Libraries

In [2]:
import pandas as pd
import numpy as np  
from sklearn.preprocessing import StandardScaler
import joblib
import os

## Loading dataset and transformations

In [24]:
df = pd.read_csv('../data/raw/LUSCexpfile.csv', delimiter=';')
df.head()

C:\Users\javie\AppData\Local\Temp\ipykernel_31500\2717525265.py:1: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,25

Unnamed: 0 03611591-08da-45c5-bfe5-ce511a2cc341  \
0        NaN                               normal   
1       A1BG                          0.108403483   
2   A1BG-AS1                          1.187325638   
3       A1CF                          0.001370356   
4        A2M                          459.7836081   

  045e2535-012c-4cec-b386-9d6f62b91829 118933b6-86c7-4668-9973-424ece352a60  \
0                               normal                               normal   
1                          0.095187869                          0.097042537   
2                          0.594016407                          1.012277195   
3                          0.032045614                                    0   
4                          648.0057316                          444.8724975   

  153cb1b1-774c-4e65-b209-6a462b68cfc5 18f0b34c-d2f6-4746-b450-8ed3def9e334  \
0                               normal                               normal   
1                           0.04302861                          0.095016254   
2                          1.015939279                          1.174120262   
3                                    0                                    0   
4                          751.7312734                          1385.150034   

  1ea50a97-2adb-4252-aec8-2fe94d49b0a0 29b6e624-9539-4abc-b80f-b5c7b65948fc  \
0                               normal                               normal   
1                          0.184691931                          0.049553602   
2                          1.830951652                           0.75448563   
3                          0.009064274                          0.004770426   
4                          656.3835066                          1068.335251   

  306df79f-3f2c-4631-b527-7d79336bc573 378d4b61-b44c-4aed-86e9-be63913aaf66  \
0                               normal                               normal   
1                          0.084045631                          0.055256044   
2                          1.125102086                          0.729134459   
3                          0.001460858                                    0   
4                          416.7646218                           793.907885   

   ... fb97b7ac-e36e-4b49-aa22-1ed7b5a0dd4b  \
0  ...                                tumor   
1  ...                          0.212273128   
2  ...                          1.017353161   
3  ...                                    0   
4  ...                          29.70095721   

  fc25b410-f91e-4607-a851-f434ff628052 fcb5bf8a-5c16-48cd-84a4-2a87695eff2b  \
0                                tumor                                tumor   
1                          0.040941764                          0.215710235   
2                          0.304290737                          1.436468418   
3                          0.007763327                          0.089985963   
4                          60.37668747                          210.0103015   

  fe819208-ff01-49e0-a04a-5fb467351305 fe8a279b-cfbb-4a96-9e30-5589bddb7911  \
0                                tumor                                tumor   
1                          0.214627375                          0.013129838   
2                          0.615680103                           0.16948915   
3                          0.007308917                          0.010954521   
4                          8.883384349                          46.51221702   

  feb82116-b8c0-41e6-9afd-43cc6f6c0d62 ff47bedc-e4c7-4775-8e8e-ed4d2eecda83  \
0                                tumor                                tumor   
1                          0.072578267                          0.210407375   
2                          1.145089115                          1.354173806   
3                           0.00137622                          0.002250611   
4                          53.39539897                          22.30662803   

  ff915b5b-98f0-4cd9-8afc-90bc16529cd8 ffae8b64-70e0-4d15-819f-7d8dd0a51db0

In [25]:
df = df.T
df.columns = df.iloc[0]
df = df[1:]
df = df.rename(columns = {np.nan:'Condition'})
df.columns.name = None

In [26]:
df.shape

(551, 56908)


Firstly convery dtypes into numpy.float64 except "Condition" column, that is our binary decision variable

In [27]:
df = df.apply(lambda x: x.astype(np.float64) if x.name != 'Condition' else x.astype('str'))
df.dtypes

Condition     object
A1BG         float64
A1BG-AS1     float64
A1CF         float64
A2M          float64
              ...   
ZYG11AP1     float64
ZYG11B       float64
ZYX          float64
ZYXP1        float64
ZZEF1        float64
Length: 56908, dtype: object

Check that only condition column is defined as object

In [22]:
df.dtypes.value_counts()

float64    551
Name: count, dtype: int64

In [28]:
df.head()

Condition      A1BG  A1BG-AS1      A1CF  \
03611591-08da-45c5-bfe5-ce511a2cc341    normal  0.108403  1.187326  0.001370   
045e2535-012c-4cec-b386-9d6f62b91829    normal  0.095188  0.594016  0.032046   
118933b6-86c7-4668-9973-424ece352a60    normal  0.097043  1.012277  0.000000   
153cb1b1-774c-4e65-b209-6a462b68cfc5    normal  0.043029  1.015939  0.000000   
18f0b34c-d2f6-4746-b450-8ed3def9e334    normal  0.095016  1.174120  0.000000   

                                              A2M   A2M-AS1     A2ML1  \
03611591-08da-45c5-bfe5-ce511a2cc341   459.783608  0.486278  0.027029   
045e2535-012c-4cec-b386-9d6f62b91829   648.005732  1.080390  0.055879   
118933b6-86c7-4668-9973-424ece352a60   444.872498  1.047981  0.012098   
153cb1b1-774c-4e65-b209-6a462b68cfc5   751.731273  0.780322  0.021787   
18f0b34c-d2f6-4746-b450-8ed3def9e334  1385.150034  0.721305  0.048111   

                                      A2ML1-AS1  A2ML1-AS2     A2MP1  ...  \
03611591-08da-45c5-bfe5-ce511a2cc341   0.000000   0.000000  0.084773  ...   
045e2535-012c-4cec-b386-9d6f62b91829   0.029601   0.000000  0.040223  ...   
118933b6-86c7-4668-9973-424ece352a60   0.000000   0.000000  0.178337  ...   
153cb1b1-774c-4e65-b209-6a462b68cfc5   0.029335   0.034351  0.082569  ...   
18f0b34c-d2f6-4746-b450-8ed3def9e334   0.000000   0.000000  0.182330  ...   

                                         ZWINT      ZXDA      ZXDB      ZXDC  \
03611591-08da-45c5-bfe5-ce511a2cc341  1.665693  0.860664  2.226001  5.013749   
045e2535-012c-4cec-b386-9d6f62b91829  1.358792  1.232901  3.272293  5.482293   
118933b6-86c7-4668-9973-424ece352a60  2.645383  0.726070  1.606960  2.783768   
153cb1b1-774c-4e65-b209-6a462b68cfc5  2.286109  0.844632  2.436370  4.161483   
18f0b34c-d2f6-4746-b450-8ed3def9e334  2.400978  0.854256  2.225530  4.401694   

                                        ZYG11A  ZYG11AP1    ZYG11B        ZYX  \
03611591-08da-45c5-bfe5-ce511a2cc341  0.008410       0.0  4.194418  99.369159   
045e2535-012c-4cec-b386-9d6f62b91829  0.025654       0.0  4.662835  98.084795   
118933b6-86c7-4668-9973-424ece352a60  0.037645       0.0  4.619178  54.605189   
153cb1b1-774c-4e65-b209-6a462b68cfc5  0.042371       0.0  4.814995  86.273891   
18f0b34c-d2f6-4746-b450-8ed3def9e334  0.006238       0.0  5.432338  93.666207   

                                      ZYXP1     ZZEF1  
03611591-08da-45c5-bfe5-ce511a2cc341    0.0  4.552543  
045e2535-012c-4cec-b386-9d6f62b91829    0.0  5.626856  
118933b6-86c7-4668-9973-424ece352a60    0.0  2.309259  
153cb1b1-774c-4e65-b209-6a462b68cfc5    0.0  5.302771  
18f0b34c-d2f6-4746-b450-8ed3def9e334    0.0  5.399341  

[5 rows x 56908 columns]

## Preprocessing

We are starting with removing non-gene columns by dropping Sample ID and separate condition as they are not predictors

In [ ]:
X_raw = df.drop(columns=["Condition"])
X_raw = X_raw.reset_index(drop=True)

y = df['Condition'].map({'tumor': 1, 'normal': 0}) # This is going to be our binary target

print(X_raw.shape)
print(y.shape)

(551, 56907)
(551,)


In [ ]:
print(y)

03611591-08da-45c5-bfe5-ce511a2cc341    0
045e2535-012c-4cec-b386-9d6f62b91829    0
118933b6-86c7-4668-9973-424ece352a60    0
153cb1b1-774c-4e65-b209-6a462b68cfc5    0
18f0b34c-d2f6-4746-b450-8ed3def9e334    0
                                       ..
feb82116-b8c0-41e6-9afd-43cc6f6c0d62    1
ff47bedc-e4c7-4775-8e8e-ed4d2eecda83    1
ff915b5b-98f0-4cd9-8afc-90bc16529cd8    1
ffae8b64-70e0-4d15-819f-7d8dd0a51db0    1
ffb473b7-a5cf-4607-b97c-fc78b4719ccb    1
Name: Condition, Length: 551, dtype: int64


#### Filter Low-Variance genes

Many genes have near-zero expression accross all samples. We are going to drop genes with very low variance too as they will not help prediction and slow down Lasso. 

In this approach we are going to keep genes with CMP > 1 in >= 10% of the samples. If not, we can approximate CMP from raw counts

In [ ]:
cpm = X_raw.div(X_raw.sum(axis=1), axis=0) * 1e6 # This is Counts Per Million Normalization
mask = (cpm > 1).sum(axis=0) >= (0.1 * X_raw.shape[0])
X_filtered = X_raw.loc[:, mask] # Takes only the genes (columns) that have more than 1 CPM in at least 10% of the samples

print(X_filtered.shape)

(551, 24763)


We can see that we reduced the number of columns from 56907 to 24763, which is a great advance and makes things easier for Lasso

#### Normalize expression

We are going to use log2(CPM + 1) for stability and cross-study consistency. This is the base 2 logarithm of the expression values plus one, this one is added to avoid issues taking the log of zero as many samples contemplate zero expression

In [ ]:
X_log2cmp = np.log2(X_filtered + 1) # Adding one to avoid errors when taking log2 of zero

print(X_log2cmp.head())

   A1BG-AS1        A2M   A2M-AS1     A2ML1    A4GALT      AAAS      AACS  \
0  1.129168   8.847946  0.571704  0.038477  3.145257  2.979703  1.521052   
1  0.672666   9.342087  1.056854  0.078444  3.154410  2.841181  1.483861   
2  1.008829   8.800487  1.034202  0.017349  3.345186  2.835205  1.283291   
3  1.011452   9.555991  0.832139  0.031095  3.648796  2.902867  1.627301   
4  1.120432  10.436868  0.783502  0.067792  3.352290  2.741777  1.507383   

      AADAC   AADACL2  AADACL2-AS1  ...      ZW10    ZWILCH     ZWINT  \
0  3.085262  0.011188     0.221307  ...  2.526893  1.184472  1.414510   
1  2.046968  0.000000     0.044892  ...  2.062550  1.152093  1.238048   
2  3.769611  0.059100     0.474656  ...  2.343435  1.218147  1.866070   
3  2.348768  0.084254     0.000000  ...  2.307015  1.190593  1.716380   
4  1.396432  0.057163     0.871823  ...  2.662970  1.431225  1.765950   

       ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1  
0  0.895818  1.689747  2.58826

#### Save gene names

We are going to save the genes names into a list so we know which position they occupay later when we standardize as the dataframe form is going to be lost

In [33]:
gene_names = X_log2cmp.columns.to_list()

print(gene_names)

['A1BG-AS1', 'A2M', 'A2M-AS1', 'A2ML1', 'A4GALT', 'AAAS', 'AACS', 'AADAC', 'AADACL2', 'AADACL2-AS1', 'AADACP1', 'AADAT', 'AAED1', 'AAGAB', 'AAK1', 'AAMDC', 'AAMP', 'AANAT', 'AAR2', 'AARD', 'AARS', 'AARS2', 'AARSD1', 'AASDH', 'AASDHPPT', 'AASS', 'AATBC', 'AATF', 'AATK', 'ABALON', 'ABAT', 'ABCA1', 'ABCA10', 'ABCA11P', 'ABCA12', 'ABCA13', 'ABCA17P', 'ABCA2', 'ABCA3', 'ABCA4', 'ABCA5', 'ABCA6', 'ABCA7', 'ABCA8', 'ABCA9', 'ABCA9-AS1', 'ABCB1', 'ABCB10', 'ABCB6', 'ABCB7', 'ABCB8', 'ABCB9', 'ABCC1', 'ABCC10', 'ABCC2', 'ABCC3', 'ABCC4', 'ABCC5', 'ABCC5-AS1', 'ABCC6', 'ABCC6P1', 'ABCC6P2', 'ABCC8', 'ABCC9', 'ABCD1', 'ABCD2', 'ABCD3', 'ABCD4', 'ABCE1', 'ABCF1', 'ABCF2', 'ABCF3', 'ABCG1', 'ABCG2', 'ABHD1', 'ABHD10', 'ABHD11', 'ABHD11-AS1', 'ABHD12', 'ABHD12B', 'ABHD13', 'ABHD14A', 'ABHD14B', 'ABHD15', 'ABHD16A', 'ABHD17A', 'ABHD17AP4', 'ABHD17AP5', 'ABHD17B', 'ABHD17C', 'ABHD18', 'ABHD2', 'ABHD3', 'ABHD4', 'ABHD5', 'ABHD6', 'ABHD8', 'ABI1', 'ABI2', 'ABI3', 'ABI3BP', 'ABL1', 'ABL2', 'ABLIM1', 'ABL

#### Standardize the Data

Lasso is very sensitive to scale as it uses L1 penalty and this is very sensitive to the scale of features. Moreover, this way we reasure the equal contribution of each gene expression. Me must achieve a consistent range and scale through dataset. All features should have mean zero and variance one as normal distribution. 

In [ ]:
scaler = StandardScaler() # x new = (x - mean) / std    
X_scaled = scaler.fit_transform(X_log2cmp)
  
print(X_scaled.shape)

print(X_scaled.mean(axis=0))  
print(X_scaled.std(axis=0))

(551, 24763)
[-1.25731246e-16 -4.12656398e-16  2.57910249e-17 ...  1.09611856e-15
 -8.38208309e-16  2.32119224e-16]
[1. 1. 1. ... 1. 1. 1.]


We can see that we achieved a 551 x 24763 numpy array storing the gene expression features; where the features have zero mean and variance equal to one

#### How all connects

So now we achieved:

- X_scaled , that is the standardized gene expression features with variance and scale reduction stored in a multidimensional numpy array

- y , our binary target for normal or tumor samples where 0 is normal and 1 is tumor stored in a one dimensional array

- gene_names , a list that contains the name of the genes

Now, are going to save them to another folder to maintain workflow

In [35]:
np.save('../data/processed/X_preprocess.npy', np.array(X_scaled))
np.save("../data/processed/y.npy", y)

joblib.dump(gene_names, "../data/processed/gene_names.pkl")

['../data/processed/gene_names.pkl']